### PR TITLE
Automatically create symbolic-only environments from PDDL

### PR DESCRIPTION
**Pull request by tomsilver**
Includes the ability to generate environments from a fixed set of PDDL problem files (as in PDDLGym), and also the ability to generate environments via procedural problem generation. There is a Blocks example for both cases.

Examples of things that work (perfect test performance):
* `python src/main.py --env pddl_blocks_fixed_tasks --approach oracle --seed 0 --num_train_tasks 5 --num_test_tasks 5`
* `python src/main.py --env pddl_blocks_fixed_tasks --approach nsrt_learning --seed 0 --num_train_tasks 5 --num_test_tasks 5`
* `python src/main.py --env pddl_blocks_fixed_tasks --approach grammar_search_invention --seed 0 --num_train_tasks 5 --num_test_tasks 5`
* `python src/main.py --env pddl_blocks_procedural_tasks --approach oracle --seed 0`
* `python src/main.py --env pddl_blocks_procedural_tasks --approach nsrt_learning --seed 0`
* `python src/main.py --env pddl_blocks_procedural_tasks --approach grammar_search_invention --seed 0`

Examples of things that run and don't crash:
* `python src/main.py --env pddl_blocks_fixed_tasks --approach random_actions --seed 0 --num_train_tasks 5 --num_test_tasks 5`
* `python src/main.py --env pddl_blocks_fixed_tasks --approach random_options --seed 0 --num_train_tasks 5 --num_test_tasks 5`
* `python src/main.py --env pddl_blocks_fixed_tasks --approach grammar_search_invention --seed 0 --num_train_tasks 5 --num_test_tasks 5 --excluded_predicates all`
*  `python src/main.py --env pddl_blocks_procedural_tasks --approach random_actions --seed 0`
* `python src/main.py --env pddl_blocks_procedural_tasks --approach random_options --seed 0 `
*  `python src/main.py --env pddl_blocks_procedural_tasks --approach grammar_search_invention --seed 0 --excluded_predicates all`

### CODE REVIEW

File: src/envs/pddl_procedural_generation.py
Comment (ronuchit) [inline src/envs/pddl_procedural_generation.py:1]: blank line after?

File: src/structs.py
Comment (ronuchit) [inline src/structs.py:1322]: strings

File: src/envs/pddl_procedural_generation.py
Comment (ronuchit) [inline src/envs/pddl_procedural_generation.py:24]: assert max_num_blocks_goal <= min_num_blocks?

File: src/envs/pddl_procedural_generation.py
Comment (ronuchit) [inline src/envs/pddl_procedural_generation.py:53]: to the most recently created pile

File: src/ground_truth_nsrts.py
Comment (ronuchit) [inline src/ground_truth_nsrts.py:2117]: before casting, should we assert isinstance(env_base, _PDDLEnv)?

File: src/envs/pddl_env.py
Comment (ronuchit) [inline src/envs/pddl_env.py:41]: fixed-dimensional

File: src/envs/pddl_env.py
Comment (ronuchit) [inline src/envs/pddl_env.py:45]: but doesn't this require knowing the number of objects? (or at least an upper bound on it)

File: src/envs/pddl_env.py
Comment (ronuchit) [inline src/envs/pddl_env.py:349]: would prefer this method to be in the base class as a staticmethod since it's so long

File: src/envs/pddl_env.py
Comment (ronuchit) [inline src/envs/pddl_env.py:118]: i feel like we should just fix this now. can't we just make this env-specific? alternatively we can just set self._train_tasks and self._test_tasks in `__init__` itself, and extract the goal predicates from the task goals. note that the comment in base_env.py "# These are generated lazily when..." doesn't matter here because we always have everything we need to be able to generate tasks in the init, for these pddl envs. in this case, _generate_train/test_tasks would be no-ops

File: src/envs/pddl_env.py
Comment (ronuchit) [inline src/envs/pddl_env.py:340]: i'm confused by this cast, don't we want to be returning a Set[GroundAtom] according to the function annotation? (i know they're the same, but we should be using one or the other). i feel like this should be casting to Set[GroundAtom] instead (or maybe you don't even need to cast...?)

Comment (ronuchit) [inline src/envs/pddl_env.py:340]: can we remove this PDDLEnvSimulatorState thing entirely? i think it's more confusing than helpful

File: src/envs/pddl_env.py
Comment (ronuchit) [inline src/envs/pddl_env.py:345]: why can't this be empty? is it because you want `list(state)` to give you the objects?

File: src/envs/pddl_env.py
Comment (ronuchit) [inline src/envs/pddl_env.py:339]: why is this out here but action_to_ground_strips_op isn't? can we just put all these methods in the base class?

File: src/envs/pddl_env.py
Comment (ronuchit) [inline src/envs/pddl_env.py:155]: parenthesize (op_arity + 1)

File: src/envs/pddl_env.py
Comment (ronuchit) [inline src/envs/pddl_env.py:168]: let's cache the domain part of this out in the init? you can make your call to _parse_pddl_domain return the pyperplan_domain in addition to current return values

File: src/envs/pddl_env.py
Comment (ronuchit) [inline src/envs/pddl_env.py:132]: let's precompute this into `self._options` in init

File: src/envs/pddl_env.py
Comment (ronuchit) [inline src/envs/pddl_env.py:82]: call this ordered_objs

File: src/envs/pddl_env.py
Comment (ronuchit) [inline src/envs/pddl_env.py:214]: tricky, add a comment to clarify what's going on in this and the next line

File: src/envs/pddl_env.py
Comment (ronuchit) [inline src/envs/pddl_env.py:218]: parenthesize (len(obj_idxs) + 1)

File: src/envs/pddl_env.py
Comment (ronuchit) [inline src/envs/pddl_env.py:235]: environment

File: src/envs/pddl_env.py
Comment (ronuchit) [inline src/envs/pddl_env.py:246]: Return a list of indices corresponding to training problems. For each idx in the returned list, it is expected that there will be a problem file named task{idx}.pddl.

File: src/envs/pddl_env.py
Comment (ronuchit) [inline src/envs/pddl_env.py:251]: Return a list of indices corresponding to test problems. For each idx in the returned list, it is expected that there will be a problem file named task{idx}.pddl.

File: src/envs/pddl_env.py
Comment (ronuchit) [inline src/envs/pddl_env.py:273]: isn't this abstract too?

File: src/envs/pddl_env.py
Comment (ronuchit) [inline src/envs/pddl_env.py:440]: assert len(problems) >= num

File: src/envs/pddl_env.py
Comment (ronuchit) [inline src/envs/pddl_env.py:297]: instead of magic numbers 6 and 11, let's grab the number of problems in the folder and use half for train, half for test?

File: src/envs/pddl_env.py
Comment (ronuchit) [inline src/envs/pddl_env.py:345]: can we make this zero-dimensional?

File: src/envs/pddl_env.py
Comment (ronuchit) [inline src/envs/pddl_env.py:372]: sigh yeah. i can't think of a way around this either...

File: src/envs/pddl_env.py
Comment (ronuchit) [inline src/envs/pddl_env.py:389]: rename to `param_name_to_param`

**Review by ronuchit — COMMENTED** [review]
looking great!!

File: src/envs/pddl_env.py
Comment (tomsilver) [reply src/envs/pddl_env.py:45]: it only requires knowing the max operator arity

File: src/envs/pddl_env.py
Comment (tomsilver) [reply src/envs/pddl_env.py:349]: I don't care either way, but what's the reason to prefer staticmethods for longer things?

Comment (ronuchit) [reply src/envs/pddl_env.py:349]: i was thinking all these methods should go into the base class... that way they're only accessible to subclasses

File: src/envs/pddl_env.py
Comment (ronuchit) [reply src/envs/pddl_env.py:45]: oh yeah that's cuz your action space has infs on the upper bound

File: src/ground_truth_nsrts.py
Comment (tomsilver) [reply src/ground_truth_nsrts.py:2117]: ignoring because of the weirdness with isinstance

File: src/envs/pddl_env.py
Comment (tomsilver) [reply src/envs/pddl_env.py:349]: discussed on slack, ignoring

File: src/envs/pddl_env.py
Comment (tomsilver) [reply src/envs/pddl_env.py:345]: yeah that's why

File: src/envs/pddl_env.py
Comment (tomsilver) [reply src/envs/pddl_env.py:273]: yeah. I changed "Concrete" to "Specific" in the heading. I still think it's good to group all the blocks stuff together

File: src/envs/pddl_env.py
Comment (tomsilver) [reply src/envs/pddl_env.py:297]: I prefer this way because I want it to be easy to experiment with different train and test problem sets. Manipulating the files themselves is a pain.

Comment (ronuchit) [reply src/envs/pddl_env.py:297]: what do you mean by manipulating the files? wouldn't it just be `len(os.listdir(path to blocks folder))`?

Comment (tomsilver) [reply src/envs/pddl_env.py:297]: I mean in the course of doing research, if I want to experiment with a certain subset of all the PDDL files I have, I don't want to have to change the files in my directory manually, I would rather change the numbers here

File: src/envs/pddl_env.py
Comment (ronuchit) [inline src/envs/pddl_env.py:457]: strings*

File: src/envs/pddl_procedural_generation.py
Comment (ronuchit) [inline src/envs/pddl_procedural_generation.py:1]: strings*

File: src/envs/pddl_env.py
Comment (ronuchit) [inline src/envs/pddl_env.py:235]: fine with me, but let's at least move this into settings.py? `pddl_blocks_fixed_train_indices` and `pddl_blocks_fixed_test_indices` for instance

**Review by ronuchit — APPROVED** [review]
lgtm after testing

**Comment by tomsilver** [comment]
@ryangpeixu this pull request is with your project in mind! I will explain in person

File: tests/envs/test_pddl_procedural_generation.py
Comment (ronuchit) [inline tests/envs/test_pddl_procedural_generation.py:72]: why spaces here but not after ontable above? can we add spaces there too?

File: src/envs/pddl_env.py
Comment (ronuchit) [inline src/envs/pddl_env.py:51]: so we're not using the base copy anymore?

**Review by ronuchit — APPROVED** [review]
lgtm again. since we are splitting test_pddl_env.py and test_pddl_blocks_env.py into separate files, do we also need to split pddl_env.py into a new file pddl_blocks_env.py? or not worth?

File: src/envs/pddl_env.py
Comment (tomsilver) [reply src/envs/pddl_env.py:51]: that's right

**Comment by tomsilver** [comment]
good point, I just consolidated the two test files for consistency with the rest of the test file structure

**Pull request merged**
